### PR TITLE
ENH: testing: argument `err_msg` of assertion functions can be any object

### PR DESCRIPTION
**Pull request by mdhaber**
Supersedes gh-17011

gh-17011 proposed calling `str` automatically on the `err_msg` argument of assertion functions to make them more tolerant and mimic the behavior of Python `assert` statements. 

It looks like the PR was inactive, so I took the liberty of merging main (to preserve @asmeurer 's) commits, making the [suggested change ](https://github.com/numpy/numpy/pull/17011#discussion_r467255315), and making a patch to ensure that this works for all assertion functions that accept `err_msg`.

<details>
For local testing:

```python3
import numpy as np

assertions = [np.testing.assert_allclose,
              np.testing.assert_equal,
              np.testing.assert_array_equal,
              np.testing.assert_array_less,
              np.testing.assert_array_almost_equal,
              np.testing.assert_almost_equal,
              np.testing.assert_approx_equal]
x = 3
y = 2
err_msg = np.array([1, 2, 3])

for assertion in assertions:
    try:
        assertion(x, y, err_msg=err_msg)
    except AssertionError as e:
        print(e)
```

I could add this as a unit test, if desired.

</details>

@mattip it [looks like](https://github.com/numpy/numpy/pull/17011#issuecomment-1064265699) you may have been willing to review the original PR. Would you be able to take a look at this one?

### CODE REVIEW

File: doc/release/upcoming_changes/17011.improvement.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Call ``str`` automatically on third argument to functions like `assert_equal`
```
Comment (mdhaber) [inline doc/release/upcoming_changes/17011.improvement.rst:1]: Oops, it looks like this should be named `24877.improvement.rst`

**Comment by asmeurer** [comment]
Cool thanks

**Review by tylerjereddy — COMMENTED** [review]
In case it helps, this test fails on `main` and passes here:

```diff
--- a/numpy/testing/tests/test_utils.py
+++ b/numpy/testing/tests/test_utils.py
@@ -1648,3 +1648,12 @@ def __del__(self):
         finally:
             # make sure that we stop creating reference cycles
             ReferenceCycleInDel.make_cycle = False
+
+@pytest.mark.parametrize("assert_func", [
+    assert_equal,
+    assert_allclose,
+    assert_array_equal])
+def test_err_msg_obj(assert_func):
+    a = np.zeros(10)
+    with pytest.raises(AssertionError):
+        assert_func(a, a + 7, err_msg=1)
```

**Comment by mdhaber** [comment]
Thanks @tylerjereddy. I'd be happy to add that and/or the one in https://github.com/numpy/numpy/pull/24877#issue-1931361537 (Details) as a unit test if desired. (I would have done that if I were starting from scratch, but It looked like the original PR was ready to merge except for a CI glitch, so I didn't think I should add more code to review if it wasn't considered necessary.)

**Review by rgommers — APPROVED** [review]
This LGTM, and all comments on the original PR were addressed. No need for a test I think - let's give this a go. Thanks @mdhaber  & all!

**Pull request merged**